src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -2676,21 +2676,38 @@ private void ReportCantConvertLambdaReturn(SyntaxNode syntax, DiagnosticBag diag
             var lambda = this.ContainingMemberOrLambda as LambdaSymbol;
             if ((object)lambda != null)
             {
+                Location location = getLocationForDiagnostics(syntax);
                 if (IsInAsyncMethod())
                 {
                     // Cannot convert async {0} to intended delegate type. An async {0} may return void, Task or Task<T>, none of which are convertible to '{1}'.
                     Error(diagnostics, ErrorCode.ERR_CantConvAsyncAnonFuncReturns,
-                        syntax,
+                        location,
                         lambda.MessageID.Localize(), lambda.ReturnType);
                 }
                 else
                 {
                     // Cannot convert {0} to intended delegate type because some of the return types in the block are not implicitly convertible to the delegate return type
                     Error(diagnostics, ErrorCode.ERR_CantConvAnonMethReturns,
-                        syntax,
+                        location,
                         lambda.MessageID.Localize());
                 }
             }
+
+            Location getLocationForDiagnostics(SyntaxNode node)
+            {
+                if (node is LambdaExpressionSyntax lambdaSyntax)
```

Comment:
You could write this `switch (node) { case LambdaExpressionSyntax lambdaSyntax: return ...`